### PR TITLE
Add registry api proxy in EdgeHub

### PR DESCRIPTION
**Pull request by philipktlin**
Add registry api proxy in EdgeHub, which is calling from iotedged on child device.  Currently it supports Get, Post, Put, Delete of registry API endpoint (e.g. /deivces/{deviceId}/modules/{moduleId}).  Add unit test and manually tested with postman.

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            HttpRequest request = context.Request;
+
+            // This pattern is used to match registry API for create, update, get, delete of a module, and list modules.
+            Regex registryApiUriPattern = new Regex(@"\/devices\/.+\/modules(\/.+)?$");
+
+            if (context.WebSockets.IsWebSocketRequest ||
+                !request.IsHttps ||
+                !registryApiUriPattern.IsMatch(request.Path))
+            {
+                return null;
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:111]: Don't use null pls.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+            await responseMessage.Content.CopyToAsync(response.Body);
+        }
+
+        internal Uri BuildDestinationUri(HttpContext context)
+        {
+            HttpRequest request = context.Request;
+
+            // This pattern is used to match registry API for create, update, get, delete of a module, and list modules.
+            Regex registryApiUriPattern = new Regex(@"\/devices\/.+\/modules(\/.+)?$");
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:105]: What is the pattern here? 
just /devices/deviceId/modules/moduleId ?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Startup.cs [hunk 3]
```diff
@@ -93,6 +81,28 @@ public void Configure(IApplicationBuilder app)
             });
         }
 
+        private (string iotHubHostname, string edgeDeviceId, string gatewayHostname) GetStartupParameters()
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Startup.cs:84]: Remove private.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs (hunk 2 above)
Comment (philipktlin) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:105]: You can find all references of iotedged calling to registry api in this file, https://github.com/Azure/iotedge/blob/4425533aed41cd3eed4780612af214a075b51093/edgelet/iothubservice/src/device.rs#L91.  This pattern is for GET, PUT, DELETE of /devices/{deviceId}/modules/{moduleId} and GET of /devices/{deviceId}/modules.

[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:105]: can't make it as static, it use class field in line 115.  Or you want me to move this class field as method argument and make it static.  Btw asp.net will use single instance of middleware to process multiple request, do we still have benefit to make it static?  Let me know if you still prefer to use static here.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs (hunk 1 above)
Comment (philipktlin) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:111]: Option is used.

**Comment by philipktlin** [comment]
> Changes look okay. But how will it work once we want to support on-behalf of authentication from EdgeHub.

Let's chat about this offline.  I would like to understand the details.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/RegistryProxyMiddleware.cs:106]: I am talking about the Regex.. that can be static right?

**Pull request merged**